Updated DeepTauv2p1 UL SFs

bfonta opened the pull request:
This PR implements the updated DeepTauv2p1 ultra-legacy scale factors (as detailed [here](https://twiki.cern.ch/twiki/bin/viewauth/CMS/TauIDRecommendationForRun2#Corrections_for_the_DeepTauv2p1)). The scale factors depend on both decay mode and (continuous) transverse momentum (the previous version depended only on binned transverse momentum), and includes a larger number of systematics (12 in total).

- section of the skimmer responsible for calculating and setting the tau ID SFs, such that there is much less code duplication across channels. 
- modifications were propagated to the three independent skimmers (2016, 2017 and 2018).
- old deepTau scale factors were kept; they might be useful for comparisons during validation
- MVA scale factors were removed
- a (very minor) bug was found and fixed: for high pT taus, the wrong tau was being cut upon when applying a cut at 100GeV. This will have no impact now, since all the systematics are now propagated to the next analysis steps (the high pT uncertainty was being so far added to the "standard" up and down variations directly in the skimmer).

----------------------------

- Validation has been performed. Results are compatible with previous weights. Some examples for the ```etau``` channel (similar for the others):

<img src="https://github.com/LLRCMS/KLUBAnalysis/assets/20703947/df7d5e68-38f5-49e7-8531-831b34e9a470" width="350">
<img src="https://github.com/LLRCMS/KLUBAnalysis/assets/20703947/1dc940a9-7520-416d-9247-7471c1e86a35" width="350">
<img src="https://github.com/LLRCMS/KLUBAnalysis/assets/20703947/79820394-479e-498e-8309-a27529773232" width="350">
<img src="https://github.com/LLRCMS/KLUBAnalysis/assets/20703947/58cac1a7-47d8-4257-aac4-82c04e8ad845" width="350">

- The script ```scripts/compareRatios.py``` was included to perform the ratios above.

- For convenience, the selection was also updated to reflect synchronization efforts with Pisa.

- On an unrelated note, the DY sample from 0 to 50 GeV was added back, and the stitching was updated as follows (where ```const float stitchWeights [2] = {1./2., 1./3.};```:

```c++
float stitchWeight = 1.;
if (DY_tostitch)
{
  float Zpt = theBigTree.lheVPt;
  if (Zpt==0.0)
	stitchWeight = stitchWeights[0];
  else
	stitchWeight = stitchWeights[1];
}
```

- *Update*: the ttbar cross-sections were updated, reflecting the new published value of the inclusive ttbar cross-section [here](https://arxiv.org/pdf/2108.02803.pdf). Comparisons with older values can be seen [here](https://twiki.cern.ch/twiki/pub/CMSPublic/PhysicsResultsTOPSummaryFigures/tt_xsec_cms_13TeV.pdf).

dzuolo commented:
Hi Bruno! First of all thanks for being dealing with this huge mess created by the tau POG :) Just a heads up while you are still working on it. I have the impression that we have to move to the VVLoose WP of DeepTauVsEle in order to use the scale factors. If that's the case we need to modify the selections in offlineProducerHelper accordingly.

portalesHEP commented on test/skimNtuple2018_HHbtag.cpp:
```diff
@@ -2831,591 +2843,477 @@ int main (int argc, char** argv)
+		  if(pType == 0) {
+			idSF_leg2_MVA_vsEle  = MVA_antiEle_vloose ->getSFvsEta(leg2eta, tau2Genmatch);
+			idSF_leg2_MVA_vsMu   = MVA_antiMu_tight   ->getSFvsEta(leg2eta, tau2Genmatch);
+			idSF_leg2_deep_vsEle = Deep_antiEle_vloose->getSFvsEta(leg2eta, tau2Genmatch);
```
shouldn't it be `vvloose` here too?

portalesHEP commented on test/skimNtuple2018_HHbtag.cpp:
```diff
@@ -2831,591 +2843,477 @@ int main (int argc, char** argv)
+		  else if(pType == 1) {
+			idSF_leg2_MVA_vsEle  = MVA_antiEle_tight  ->getSFvsEta(leg2eta, tau2Genmatch);
+			idSF_leg2_MVA_vsMu   = MVA_antiMu_loose   ->getSFvsEta(leg2eta, tau2Genmatch);
+			idSF_leg2_deep_vsEle = Deep_antiEle_vloose->getSFvsEta(leg2eta, tau2Genmatch);
```
same?

bfonta commented:
The new commits brings the following:
- All vsElectrons wre changed to *VVLoose" and all vsMuons to "Tight"
- the removal of all MVA legacy Tau Id variables (all SFs and ```tauAntiMuIdx``` and ```tauAntiEleIdx```). @dzuolo Could you please double check those are indeed not required?
- bugfix: add missing variables in tautau channel

dzuolo commented:
Hi @bfonta! 
The MVA-based tau ID is superseeded by DeepTau so there is no need to keep it in our ntuples, thanks for removing it. 
Did you check with valeria and konstantin if they also use the Tight WP of DeepTauVsMu? i do not remember if that is the case honestly. 
Also, I might need to review he changes a bit better but could you point to the new branches for the Isolation scale factors?

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,455 @@ int main (int argc, char** argv)
+		idSF_leg2_deep_vsJet_pt = Deep_antiJet_medium_pt->getSFvsPT(leg2pt, tau2Genmatch);
+		idSF_leg2_deep_vsJet_2d = Deep_antiJet_2d->getSFvsDMandPT(leg2pt, tau2DM, tau2Genmatch);
+
+		if(pType == 0) {
```
Suggestion: if we end up applying the same WPs in every channel this if-else statement is not necessary.

bfonta replied:
I did not realize, thanks!

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,455 @@ int main (int argc, char** argv)
+		for (int bin = 0; bin < (int) isthisEta_IDbin_second.size(); bin++) {
+		  if (isthisEta_IDbin_second[bin])
+			{
+			  if (pType==2) { //loose working point only for TauTau
```
Same

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,455 @@ int main (int argc, char** argv)
+		for (int bin = 0; bin < (int) isthisSDet_IDbin_second.size(); bin++) {
+		  if (isthisSDet_IDbin_second[bin])
+			{
+			  if (pType==2) { //very very loose working point only for TauTau
```
Same

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,455 @@ int main (int argc, char** argv)
+	
+    if (isMC and leg1eta < 2.4) {
+	  if (pType == 0 or pType == 3) {
+		idSF_leg1 = myIDandISOScaleFactor[0]->get_ScaleFactor(leg1pt, leg1eta)*myIDandISOScaleFactor[2]->get_ScaleFactor(leg1pt, leg1eta);
```
Ah, this is the electron or muon id and iso scale factor right?

bfonta replied:
Yes, hence the check whether the first leg is a muon or an electron (```pType``` == 0, 1, 3 or 4). These scale factors were kept the same, since the tau POG did not update them.

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,455 @@ int main (int argc, char** argv)
+		idSF_deep_2d	 = idSF_leg1 * idSF_leg2_deep_vsJet_2d;
+		fakeRateSF_deep	 = idSF_leg2_deep_vsEle * idSF_leg2_deep_vsMu;
+
+		except_VsJet = idSF_leg1 * idSF_leg2_deep_vsEle * idSF_leg2_deep_vsMu;
```
I think I understand after a couple more readings. This a very smart trick to have the proper combinations of central values and systematic variations, right?

bfonta replied:
"Very smart" is quite an exaggeration :)
It is just a variable that stores all values required to calculate the weights *except* one of them. And yes, it is useful for the systematics, to avoid computing the same quantities over and over again.

dzuolo commented:
> @dzuolo The muon and electrons Id SFs were not updated. As for the tight muon WP, it is the recommended one by the POG, and actually proposed by Konstantin (it is the WP that was used by the POG do derive the vsJet corrections, so should be the default one, although the POG does not expect big dependencies of vsJet with vsMuon). The new commit covers your first comment, plus some variables I forgot to add to the tree.

Hi Bruno! Indeed the electrons and muons Id and Iso SF did not change but it took me sometime to find them in the diff produced by git eheh
Thank you for your explanation about the DeepTauVsMu Tight WP, this choice will further simplify the code and I am happy about this :)

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,430 @@ int main (int argc, char** argv)
+	  {
+		idSF_leg1_deep_vsJet_dm = Deep_antiJet_medium_dm->getSFvsDM(leg1pt, tau1DM, tau1Genmatch);
+		idSF_leg1_deep_vsJet_pt = Deep_antiJet_medium_pt->getSFvsPT(leg1pt, tau1Genmatch);
+		idSF_leg2_deep_vsJet_2d = Deep_antiJet_2d->getSFvsDMandPT(leg2pt, tau2DM, tau2Genmatch);
```
I suspect this should be leg1 and not leg2, probably also in the other skimmers

bfonta replied:
Nice catch, I had forgotten to update this in 2016 and 2017.

dzuolo commented:
I went another time through the 2016 skimmers and found a couple of typos that are probably repeated also in the 2017 and 2018 skimmers.
I thought I would have never asked something like this but probably I am getting old: could you please fix the indentations of the lines you added? Few of them are quite hard to follow in the git diff.
Finally: thanks for having worked on this, it must have been quite annoying!

bfonta commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,598 +2581,430 @@ int main (int argc, char** argv)
+		// idFakeSF_tauid_pt35to40_down	 = except_VsJet * idSF_leg2_deep_vsJet_pt_down[3];
+		// idFakeSF_tauid_pt40toInf_down = except_VsJet * idSF_leg2_deep_vsJet_pt_down[4];
+
+		idFakeSF_mutauFR_etaLt0p4_up      = except_vsMu * idSF_leg2_deep_vsMu_up[1];
```
Another very nice catch, than you.

bfonta commented:
@dzuolo I've fixed my bugs; thanks a lot for looking into it, again. I've also added the old vsJet systematics just for comparison studies if needed (I had commented them out in previous commits).
Regarding the indentation, I tried to fix it in some places, but the diff is hard to follow because the current indentation is very different than the old one (which IMHO was terrible); the only way to make a good diff would be not to touch the indentation. In the future I will try to stick with the current indentation to facilitate the code review; sorry for all the trouble.

dzuolo commented:
Thank you for the fixes Bruno! And don't worry about the indentations, way or another I could read everything.

bfonta commented:
@dzuolo FYI, I've updated the selection according to what was recently discussed with you, Konstantin and Valeria.

dzuolo commented:
@bfonta ciao bruno! Can you please check if the eta selections we apply to the leptons are compatible with the scale factors we use? One of our students is producing some skimmed ntuples using the relaxed cuts (2.3 for electrons and muons) and it seems that the are no SF for eta > 2.1 ...

dzuolo commented:
Hi @bfonta ! Thanks for posting the plots here and on mattermost. It is good to see that the impact on 2018 is minimal since we were observing good agreement between data and MC even with the old scale factors.
I will try to go once more through the PR in the coming weeks.
My questions about the muon and electron eta cut remains. I will investigare during the next tuple production if for 2016 we are accepting events with a muon with eta > 2.1 that should not pass the trigger acceptance criteria since the ```Muon_Mu19leg_2016BtoH_eff``` scale factor files contains entries up to eta 2.1

dzuolo commented on resonantLimits/run_limits.py:
```diff
@@ -94,28 +94,28 @@ def run_limits(in_tags, channels, selections, selection_prefixes, masses,
+    #               )
     selection_prefixes = ('s1b1jresolvedMcut', 's2b0jresolvedMcut', 'sboostedLLMcut')
     
     #masses = ('250', '260', '280', '300', '320', '350', '400', '450', '500', '550', '600', '650', '700', '750', '800', '850', '900', '1000', '1250', '1500', '1750', '2000', '2500', '3000')
```
I do not think in this script you wanted to comment every step apart from limits, can you add back the previous version?

dzuolo commented on src/triggerReader_cross.cc:
```diff
@@ -341,7 +341,7 @@ bool triggerReader_cross::checkOREleEleNew  (Long64_t triggerbit_1, Long64_t mat
 	firedPath = _allTriggers.at(_eeTriggers.at(i));
 	boost::regex re_tau1{"Ele(\\d+)"};
-	ptCut = checkPtCutSingle(thisPath, firedPath, re_tau1, pt_tau1, 1.0); // 1.0 GeV from HTT twiki Run2 Legacy analisys
+	ptCut = checkPtCutSingle(thisPath, firedPath, re_tau1, pt_tau1, 2.0);
```
Where does this 2 GeV threshold come from?

dzuolo replied:
Ok thanks! If there is no mention in the tau trigger twiki than it's better to leave it as it is.

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -1465,15 +1381,15 @@ int main (int argc, char** argv)
       if (oph.isMuon(dauType))
       {
-	bool passMu   = oph.muBaseline (&theBigTree, idau, 19., 2.3, 0.15, OfflineProducerHelper::MuTight, string("All") , (DEBUG ? true : false));
+	bool passMu   = oph.muBaseline (&theBigTree, idau, 20., 2.4, 0.15, OfflineProducerHelper::MuTight, string("All") , (DEBUG ? true : false));
```
We need to verify that trigger and id+iso scale factors exist up to eta 2.4

bfonta replied:
[Twiki](https://twiki.cern.ch/twiki/bin/view/CMS/MuonUL2018#Trigger_efficiency_AN1)

- [X] UL Trigger SFs: exist up to 2.4 for all years ([presentation](https://indico.cern.ch/event/1080036/contributions/4542924/attachments/2318322/3947065/210928_ULTriggerSF_kHwang.pdf), [json with SFs](https://gitlab.cern.ch/cms-muonPOG/muonefficiencies/-/blob/master/Run2/UL/2018/2018_trigger/Efficiencies_muon_generalTracks_Z_Run2018_UL_SingleMuonTriggers_schemaV2.json#L38)). They include SFs for both ```HLT_IsoMu24``` and ```HLT_IsoMu27```.

I did test what happens with our current Legacy weights when we ask for a particular SF label for 2018 data (```Muon_Run2018_IsoMu20.root``` file):

```
root [10] etaBinsH->GetXaxis()->FindFixBin(2.3)
(int) 4
root [11] etaBinsH->GetXaxis()->FindFixBin(2.4)
(int) 4
root [12] etaBinsH->GetXaxis()->FindFixBin(2.5)
(int) 5
root [13] etaBinsH->GetXaxis()->FindFixBin(999999)
(int) 5
root [14] etaBinsH->GetXaxis()->GetBinLabel(3)
(const char *) "Eta1p2to2p1"
root [15] etaBinsH->GetXaxis()->GetBinLabel(4)
(const char *) "EtaGt2p1"
root [16] etaBinsH->GetXaxis()->GetBinLabel(5)
(const char *) ""
```

We currently cut on 2.4, so between 2.1 and 2.4 our muons will be assigned the weight stored in "EtaGt2p1". This is why we get no error.
For 2016 (file ```Muon_Mu19leg_2016BtoH_eff.root```), you only have three labels: "EtaLt0p9", "Eta0p9to1p2" and "Eta1p2to2p1", hence the error. 

To recap, the issue should be gone once we update to UL SFs.

- [X] UL Iso/Id SFs: the versions we currently have are already Ultra-Legacy. Both ID and ISO SFs for muons go up to 2.4, so all good!

dzuolo replied:
Thanks for checking Bruno. This means that before running the 2016 production we will have to either switch to the UL SF or temporarily revert the cut to 2.1

bfonta replied:
Yes. I suggest not touching the code (to save your time!), and I will do a PR with the new SFs asap for the three years.

dzuolo replied:
Ok, sounds good to me! Thanks a lot!

dzuolo commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -1465,15 +1381,15 @@ int main (int argc, char** argv)
       else if (oph.isElectron(dauType))
       {
-	bool passEle   = oph.eleBaseline (&theBigTree, idau, 25., 2.3, 0.1, OfflineProducerHelper::EMVATight, string("Vertex-LepID-pTMin-etaMax") , (DEBUG ? true : false));
+	bool passEle   = oph.eleBaseline (&theBigTree, idau, 20., 2.5, 0.1, OfflineProducerHelper::EMVATight, string("Vertex-LepID-pTMin-etaMax") , (DEBUG ? true : false));
```
We need to verify that trigger and id+iso scale factors exist up to eta 2.5

bfonta replied:
[Twiki](https://twiki.cern.ch/twiki/bin/viewauth/CMS/EgHLTScaleFactorMeasurements) for electrons

- [x] UL trigger SFs: From the files I inspected it seems we always have (Legacy) the "EtaGt2p1" label (including 2016), so no error should be thrown. The new UL legacy files (@portalesHEP suggested using the ones from Ka Wa Ho) go up to 2.5.

- [x] UL id/iso SFs: the versions we currently have are already Ultra-Legacy. Both ID and ISO SFs for electrons (merged into single file) go up to 2.5, so all good!

dzuolo replied:
Thanks also for checking here. The error was probably coming from muons then. In any case we should move to the UL SF ASAP.

bfonta commented on config/selectionCfg_ETau_UL18_mHH.cfg:
```diff
@@ -98,7 +98,7 @@ TTclass     = VBFloose, mpp_tt_v5
 ## NOTE: no weight is applied for data (the simple Fill() is used)
 [selectionWeights]
-baseline = MC_weight, PUReweight, L1pref_weight, trigSF, IdAndIsoAndFakeSF_deep_pt, prescaleWeight, PUjetID_SF, bTagweightReshape
+baseline = MC_weight, PUReweight, L1pref_weight, trigSF, IdSF_deep_2d, prescaleWeight, PUjetID_SF, bTagweightReshape
```
Nice catch; this was spotted a while back by @portalesHEP, and has been fixed (it actually did not change the results in any significant way).

bfonta commented on test/skimNtuple2016_HHbtag.cpp:
```diff
@@ -2668,619 +2579,451 @@ int main (int argc, char** argv)
+	  idSF_leg1_deep_vsJet_2d_systuncorrdmeras_down		= Deep_antiJet_2d->getSFvsDMandPT(leg1pt, tau1DM, tau1Genmatch, "SystUncorrDMErasDown");
+	  idSF_leg1_deep_vsJet_2d_systgt40_up				= Deep_antiJet_2d->getSFvsDMandPT(leg1pt, tau1DM, tau1Genmatch, "SystGt40Up");
+	  idSF_leg1_deep_vsJet_2d_systgt40_down				= Deep_antiJet_2d->getSFvsDMandPT(leg1pt, tau1DM, tau1Genmatch, "SystGt40Down");
 	}
```
Is this what you mean?

![Screenshot from 2023-06-26 15-02-58](https://github.com/LLRCMS/KLUBAnalysis/assets/20703947/bdfb535d-0d71-4f9a-b4d7-f2721478301f)

dzuolo replied:
Exactly this! Thanks a lot!

bfonta commented:
I've added the corrections for high pT taus which were so far missing. Validation will take some time because it requires reproducing all skims. In the meantime let me know if all look alright.

bfonta commented:
All ready for merging. Any last minute comments?

bfonta commented:
Good point: this PR was NOT tested for 2016 and 2017. However, the old weights are being kept in the skims, so checking potential issues will be trivial.
